### ISSUE TITLE
A question about actors, tasks, and creation of workers

### ISSUE THREAD

**Issue by Vis-Viva** [ray-project/ray#7390]
Ok first, before I get told to read the docs etc, I have read them very extensively. I believe they probably _do_ answer this question, but I just need to be sure. This is because I have what appears to be a highly technical issue, but I think the solution to that issue may come down to a potential problem with my conceptual understanding of Ray that would be much easier to correct and address than the technical problem itself. So bear with me here, I'm just checking my understanding so I can maybe fix the problem myself without burdening the helpful people here with having to fix my code for me.

Also this is literally the first time I've ever posted on github or anywhere else for help with code, so pls forgive me if I'm doing anything here that's considered bad manners etc.

I have a program which does the following. The idea is that I start with a decision tree representing a game. The tree is huge, like on the order of 10^100 nodes (nodes=game states, branches=actions), so obv can't traverse the entire tree and produce a (state,action) value lookup table for every node. So, idea is, traverse part of the tree, collect (action,value) pairs and use these pairs as training data for a NN to estimate value. The NN training isn't relevant here, as all I'm working on at the moment is the tree traversal procedure. I've turned to Ray to parallelize this procedure since running a single tree traveller alone on one CPU would mean that 1) I obv can't make use of the other 95 cores I have available to me, 2) the data collection process would take 352 straight days of tree searching lol. So, here is how I'm handling that process:

![IMG_20200301_162408202](https://user-images.githubusercontent.com/47202575/75635235-04380600-5bda-11ea-9c72-6cc285711191.jpg)

Forgive me if any of my handwriting is illegible. Will clarify if needed. You can consider the Collector tasks interminable, since the tree is so gargantuan. Every time a Collector encounters a new node, it will ask the Tracker how many terminal nodes have been encountered by all Collectors thus far. If this number is ≥ some predefined number, the Collector task terminates without returning anything. Supposedly, once this condition is achieved, all Collector tasks will eventually do the same, leaving the ValueTracker actor no tasks to interact with it. At this point, the data it stores is written to a file and sent to another machine to be used for NN training, but that's beyond the scope of this question.

So, is this architecture legit? Am I thinking about this correctly? Can I have a single actor (the ValTracker) with mutable state that can be accessed/modified by some number of concurrent tasks (the Collectors)? As I get feedback on this, I may start making this question more specific and getting into the actual syntax my implementation to find where my problem lies. But for now, I just need to know whether I'm approaching this properly.

Thank you to any helpful people out there who've slogged through this pile of verbosity :)

TL;DR: Check out the picture. Is this a legit parallel tree search architecture for Ray?

**Comment by richardliaw** [comment]
I think your main question (if I understood your post correctly) is: 

>  Can I have a single actor (the ValTracker) with mutable state that can be accessed/modified by some number of concurrent tasks (the Collectors)?

If so, the answer is yes!

**Comment by Vis-Viva** [comment]
Excellent - tyvm for the quick reply. You definitely understood correctly - all the extra details are just there because you never really know where an error in your thinking about a supposedly simple question can lie. I'll post some of the syntax I'm using to implement this in a sec and I'll see if I can work toward the solution to this problem from the bottom up...

**Comment by Vis-Viva** [comment]
Ultimately the problem here seems to be that the way I've implemented this is creating a huge number of concurrent workers and eating up a whole lot of memory. Maybe given the nature of the task, this is unavoidable and I need to scale up the hardware I'm on. But first need to see if it's an error on my part. 

Specifically the problem is, I rapidly get hundreds of warnings that I'm running hundreds of workers (having started with just 4 concurrent recursive tasks), and eventually get an OutOfMemoryError. I don't have the console output handy right now, but will update this post with them when I'm back home.

Here's a bit of code from the training algorithm I'm running. I think people around here usually submit tiny toy problems to illustrate the problem they're having, but I'm a noob and it's hard for me to isolate exactly where the problem is coming from here. That said, it may look like a lot, but I think it's like half whitespace and comments. Happy to add any necessary clarifications and will probably trim some more fat when I'm home tonight (as well as give further specific details on the errors I'm running into). Thanks to any helpful people that come by here 😄 

https://gist.github.com/Vis-Viva/28e8ef01d0b9b0fca6839e8859d9414a

**Comment by Vis-Viva** [comment]
Hah, nevermind, fixed it. Turns out there were a couple of remote function calls I didn't wrap in ray.get(...) - weird that this resulted in the creation of an army of workers but ¯\_(ツ)_/¯
